### PR TITLE
DSC Module (Experimental)

### PR DESCRIPTION
**Pull request by twangboy**
Ability to compile a config using `dsc.compile_config`
Ability to apply a compiled config using `dsc.apply_config`
Ability to compile and apply a config using `dsc.run_config`

Fixes: https://github.com/saltstack/zh/issues/445
https://github.com/saltstack/zh/issues/313

### CODE REVIEW

File: salt/modules/win_dsc.py [hunk 1]
```diff
@@ -201,20 +204,174 @@ def remove(name):
...
+
+    To cache a config script to the system from the master and compile it:
+
+    .. code-block:: bash
+
+        salt '*' dsc.compile_apply_config C:\\DSC\\WebsiteConfig.ps1 salt://dsc/configs/WebsiteConfig.ps1
+    '''
+    ret = compile_config(path, source, config, salt_env)
+
+    if 'Exists' in ret and ret['Exists']:
```
Comment (cachedout) [inline salt/modules/win_dsc.py]: `if ret.get('Exists'):`

File: salt/modules/win_dsc.py [hunk 2]
```diff
@@ -226,49 +383,53 @@ def run_config(path, source=None, saltenv='base'):
...
             log.error(error)
             raise CommandExecutionError(error)
+        else:
+            config = os.path.dirname(cached_files[0])
 
     # Make sure the path exists
-    if not os.path.exists(path):
-        error = '"{0} not found.'
+    if not os.path.exists(config):
+        error = '"{0} not found.'.format(config)
```
Comment (cachedout) [inline salt/modules/win_dsc.py]: Did you want a closing double-quote here?

**Comment by damon-atkins** [comment]
See Windows PowerShell Desired State Configuration (DSC) #20548

**Comment by damon-atkins** [comment]
And also Create DSC execution module #24942

**Comment by absolutejam** [comment]
Could be a good idea to support `Invoke-DSCResource`, to run ad-hoc DSC resources? This seems to be how the freshly merged `win_dsc` does it in Ansible.

**Pull request merged**